src/DataStructures/DynamicBuffer.hpp

```diff
@@ -0,0 +1,104 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <vector>
+
+#include "DataStructures/DataVector.hpp"
+#include "Utilities/TypeTraits/IsIterable.hpp"
+
+/**
+ * \ingroup DataStructuresGroup
+ *
+ * \brief A dynamically sized vector of DataVectors or ComplexDataVectors. If T
+ * is a floating point type, `DynamicBuffer<T>` resolves to std::vector<T>.
+ *
+ * \details This class is useful when one wants to create a `std::vector<T>`
+ * with a size that is unknown at compile time. It allocates all DataVectors in
+ * a single memory chunk rather than allocating each individually. If the size
+ * of the vector is known at compile time, a `TempBuffer` object should be used
+ * instead.
+ */
+
+template <typename T>
+class DynamicBufferImpl {
+ public:
+  DynamicBufferImpl() = default;
+  DynamicBufferImpl(size_t outer_size, size_t inner_size)
+      : inner_size_(inner_size),
+        data_vectors_(outer_size),
+        buffer_(outer_size * inner_size) {
+    set_references_();
+  }
+
+  ~DynamicBufferImpl() = default;
+  DynamicBufferImpl(const DynamicBufferImpl& other) noexcept
+      : inner_size_(other.inner_size_),
+        data_vectors_(other.size()),
+        buffer_(other.buffer_) {
+    set_references_();
+  }
+  DynamicBufferImpl(DynamicBufferImpl&& other) noexcept : DynamicBufferImpl() {
+    swap(*this, other);
+  }
+  DynamicBufferImpl& operator=(DynamicBufferImpl other) noexcept {
+    swap(*this, other);
+    return *this;
+  }
+
+  inline T& operator[](size_t index) { return data_vectors_[index]; }
```

Comment:
These comments here apply generally below:
- don't need to mark anything in the body of a class as `inline`, that is implicit
- for value return types (e.g. `const_iterator`) I don't think you need to mark the resulting object as `const`. I.e. `typename std::vector<T>::const_iterator` is fine. If you want, you can also use `auto` as the return type there, or to be more careful, `decltype(auto)`. I'd personally switch all the `begin` and `end`s to return `auto` because it's fewer chars, but up to you :)